### PR TITLE
feat(textfield): Annotate textfield for Closure Compiler.

### LINKED ISSUE #345: Annotate mdc-textfield for closure

### PR DESCRIPTION
**Pull request by bwobrien**
Annotates MDC Textfield for the Closure Compiler.

Resolves #345

### CODE REVIEW

File: packages/mdc-textfield/adapter.js [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+
+  /**
+   * Returns an object representing the native text input element, with a
+   * similar API shape. The object returned should include the value, disabled
+   * and badInput properties, as well as the checkValidity() function. We never
+   * alter the value within our code, however we do update the disabled
+   * property, so if you choose to duck-type the return value for this method
+   * in your implementation it's important to keep this in mind. Also note that
+   * this method can return null, which the foundation will handle gracefully.
+   * @return {HTMLInputElement|NativeInputType}
```
Comment (acdvorak) [inline packages/mdc-textfield/adapter.js]: Add [explicit nullable operator(s)](https://google.github.io/styleguide/jsguide.html#jsdoc-nullability):

`@return {?HTMLInputElement|?NativeInputType}`
or
`@return {!HTMLInputElement|!NativeInputType|null}`
or
`@return {HTMLInputElement|NativeInputType|null}`

(I'm not sure which variation is more readable, correct, and Closure-friendly...)

File: packages/mdc-textfield/foundation.js [hunk 2]
```diff
@@ -233,17 +316,28 @@ export default class MDCTextfieldFoundation extends MDCFoundation {
...
 
+  /**
+   * @return {!HTMLInputElement|!NativeInputType} The native text input from the
+   * host environment, or a dummy if none exists.
+   * @private
+   */
   getNativeInput_() {
-    return this.adapter_.getNativeInput() || {
+    return this.adapter_.getNativeInput() ||
+    /** @type {NativeInputType} */ ({
```
Comment (acdvorak) [inline packages/mdc-textfield/foundation.js]: Add [explicit nullability operator](https://google.github.io/styleguide/jsguide.html#jsdoc-nullability):

`@type {!NativeInputType}`

File: packages/mdc-textfield/index.js [hunk 3]
```diff
@@ -103,6 +156,15 @@ export class MDCTextfield extends MDCComponent {
     };
   }
 
+  /**
+   * @return {!{
```
Comment (acdvorak) [inline packages/mdc-textfield/index.js]: Should we create `@record` definitions for these collections of methods?

Comment (kfranqueiro) [reply packages/mdc-textfield/index.js]: These subsets of the adapter API are kind of arbitrary (I assume we only split it like this as an attempt at readability); the only thing that really matters is the collective adapter API. Kind of unfortunate that we have to annotate these at all...

File: packages/mdc-textfield/adapter.js (hunk 1 above)
Comment (bwobrien) [reply packages/mdc-textfield/adapter.js]: Done. I think `null` is generally discouraged when `?` would suffice, so I went with `{?A|?B}`.

File: packages/mdc-textfield/foundation.js (hunk 2 above)
Comment (bwobrien) [reply packages/mdc-textfield/foundation.js]: Done, found some more and fixed throughout.

**Pull request merged**